store.rb

```diff
@@ -35,14 +59,25 @@ def session_id
   def stop()
     @log.warn "stopping the store"
     Process.kill("TERM", Process.pid) unless $EXIT
+    @root_watcher.unsubscribe unless @root_watcher.nil?
+    Thread.kill(@fetch_thread) unless @fetch_thread.nil?
     @zk.close() if @zk
   end
 
+  # get instances for a given service
   def nodes()
+    return load_instances_from_zk unless @use_cache
+    instances = get_cached_instances()
+    instances ||= {}
+    return instances
+  end
+
+  def load_instances_from_zk()
+    @log.info "Reading instances from zk:"
     from_server = {}
 
     begin
-      @zk.children('/').each do |child|
+      @zk.children('/', :watch => true).each do |child|
```

Comment:
I think I don't fully understand how this and the root watcher relate, but basically everyone knows more about zk than me.